File: apps/runtime/src/remote/watch.rs
//! External-change watching for editor files on an execution host (H02).
//!
//! There is no filesystem event to subscribe to across an `ssh` pipe, so this
//! is a poll and the code says so rather than pretending otherwise: one
//! `WATCH_POLL` per workspace per tick carries every path that workspace's
//! editor nodes have open, and the difference against the last answer is
//! published as the same [`WorkspaceEvent::FileChanged`] the local watcher
//! emits.
//!
//! Two honest limits, both visible to the client:
//!
//! * Latency is the poll interval, not the write.
//! * A replace in place reports `modified`. The device/inode pair the local
//!   watcher uses to tell them apart is a local detail that is deliberately
//!   not serialized, and inventing one from the remote metadata would be a
//!   guess.

use std::{
    collections::{HashMap, HashSet},
    sync::{Arc, LazyLock, Mutex},
    time::Duration,
};

use armadra_protocol::v1::WorkerServiceOperation;

use crate::{
    error::{AppError, AppResult},
    events::{EventHub, FileChangeKind, WorkspaceEvent},
    file_watch::{FileVersion, WatchRegistration, WatchStatus},
    remote::{
        client::RemoteWorker,
        service::{MAX_WATCH_PATHS, WatchPollPayload, WatchPollResult},
    },
};

/// How often a workspace with open remote editors is polled.
const INTERVAL: Duration = Duration::from_secs(2);

struct Watched {
    viewers: HashSet<String>,
    known: FileVersion,
}

struct WorkspaceWatch {
    root_path: String,
    files: HashMap<String, Watched>,
    /// Set while a poll task is running for this workspace.
    polling: bool,
}

#[derive(Default)]
struct Registry {
    workspaces: HashMap<String, WorkspaceWatch>,
}

static WATCHES: LazyLock<Mutex<Registry>> = LazyLock::new(Mutex::default);

fn registry() -> AppResult<std::sync::MutexGuard<'static, Registry>> {
    WATCHES
        .lock()
        .map_err(|_| AppError::Internal("Remote watch registry unavailable".into()))
}

/// Read one path's version on the execution host.
pub async fn version(
    worker: &RemoteWorker,
    workspace_id: &str,
    root_path: &str,
    path: &str,
) -> AppResult<FileVersion> {
    let (status, body) = worker
        .service(
            workspace_id,
            root_path,
            WorkerServiceOperation::FileVersion,
            serde_json::to_vec(&crate::remote::service::PathPayload {
                path: path.to_owned(),
            })
            .map_err(|_| AppError::Internal("Request could not be encoded".into()))?,
            false,
            false,
        )
        .await?;
    decode(status, &body)
}

/// Register one editor node's view of a remote file. Idempotent per viewer,
/// and the answer carries the version on the execution host right now.
pub async fn register(
    worker: Arc<RemoteWorker>,
    events: EventHub,
    workspace_id: &str,
    root_path: &str,
    path: &str,
    node_id: &str,
) -> AppResult<WatchRegistration> {
    if node_id.is_empty() || node_id.len() > 128 {
        return Err(AppError::BadRequest("A node id is required".into()));
    }
    let current = version(&worker, workspace_id, root_path, path).await?;
    let relative = current.path.clone();
    let start = {
        let mut registry = registry()?;
        let entry = registry
            .workspaces
            .entry(workspace_id.to_owned())
            .or_insert_with(|| WorkspaceWatch {
                root_path: root_path.to_owned(),
                files: HashMap::new(),
                polling: false,
            });
        if entry.root_path != root_path {
            entry.root_path = root_path.to_owned();
            entry.files.clear();
        }
        if entry.files.len() >= MAX_WATCH_PATHS && !entry.files.contains_key(&relative) {
            return Err(AppError::Conflict(
                "Too many watched files on this execution host".into(),
            ));
        }
        let file = entry.files.entry(relative).or_insert_with(|| Watched {
            viewers: HashSet::new(),
            known: current.clone(),
        });
        file.viewers.insert(node_id.to_owned());
        // A re-open re-baselines: what is on the host now is what the node shows.
        file.known = current.clone();
        let start = !entry.polling;
        entry.polling = true;
        start
    };
    if start {
        spawn(worker, events, workspace_id.to_owned());
    }
    Ok(WatchRegistration {
        status: WatchStatus::Watching,
        reason: None,
        version: current,
    })
}

/// Drop one viewer; the workspace stops being polled with its last file.
pub fn unregister(workspace_id: &str, path: &str, node_id: &str) -> AppResult<()> {
    let mut registry = registry()?;
    let Some(entry) = registry.workspaces.get_mut(workspace_id) else {
        return Ok(());
    };
    let relative = crate::security::workspace_relative_path(path)?;
    if let Some(file) = entry.files.get_mut(&relative) {
        file.viewers.remove(node_id);
        if file.viewers.is_empty() {
            entry.files.remove(&relative);
        }
    }
    if entry.files.is_empty() {
        registry.workspaces.remove(workspace_id);
    }
    Ok(())
}

/// Stop polling a whole workspace: read access revoked, or it went away.
pub fn release_workspace(workspace_id: &str) {
    if let Ok(mut registry) = registry() {
        registry.workspaces.remove(workspace_id);
    }
}

fn spawn(worker: Arc<RemoteWorker>, events: EventHub, workspace_id: String) {
    tokio::spawn(async move {
        loop {
            tokio::time::sleep(INTERVAL).await;
            let request = {
                let Ok(registry) = registry() else { break };
                let Some(entry) = registry.workspaces.get(&workspace_id) else {
                    break;
                };
                if entry.files.is_empty() {
                    break;
                }
                (
                    entry.root_path.clone(),
                    entry.files.keys().cloned().collect::<Vec<_>>(),
                )
            };
            let Ok(versions) = poll(&worker, &workspace_id, &request.0, request.1).await else {
                // A host that is down is not a change. The next tick retries,
                // and the connection's own backoff keeps that cheap.
                continue;
            };
            for change in difference(&workspace_id, versions) {
                if let WorkspaceEvent::FileChanged { path, .. } = &change
                    && crate::git_discovery::affects_repositories(path)
                {
                    crate::git_discovery::invalidate(&workspace_id);
                }
                events.publish(&workspace_id, change);
            }
        }
        if let Ok(mut registry) = registry()
            && let Some(entry) = registry.workspaces.get_mut(&workspace_id)
        {
            entry.polling = false;
        }
    });
}

async fn poll(
    worker: &RemoteWorker,
    workspace_id: &str,
    root_path: &str,
    paths: Vec<String>,
) -> AppResult<Vec<FileVersion>> {
    let (status, body) = worker
        .service(
            workspace_id,
            root_path,
            WorkerServiceOperation::WatchPoll,
            serde_json::to_vec(&WatchPollPayload { paths })
                .map_err(|_| AppError::Internal("Request could not be encoded".into()))?,
            false,
            false,
        )
        .await?;
    let result: WatchPollResult = decode(status, &body)?;
    Ok(result.versions)
}

/// Compare the poll against what each node last saw, updating the baseline.
fn difference(workspace_id: &str, versions: Vec<FileVersion>) -> Vec<WorkspaceEvent> {
    let Ok(mut registry) = registry() else {
        return Vec::new();
    };
    let Some(entry) = registry.workspaces.get_mut(workspace_id) else {
        return Vec::new();
    };
    let mut changes = Vec::new();
    for current in versions {
        let Some(file) = entry.files.get_mut(&current.path) else {
            continue;
        };
        if current.exists == file.known.exists && current.sha256 == file.known.sha256 {
            file.known = current;
            continue;
        }
        let kind = if !current.exists {
            FileChangeKind::Removed
        } else if !file.known.exists {
            FileChangeKind::Replaced
        } else {
            FileChangeKind::Modified
        };
        changes.push(WorkspaceEvent::FileChanged {
            workspace_id: workspace_id.to_owned(),
            path: current.path.clone(),
            kind,
            sha256: current.sha256.clone(),
            size: current.size,
            mtime: current.mtime.clone(),
        });
        file.known = current;
    }
    changes
}

/// A poll answer that is not 200 is the execution host's own error, and keeps
/// its meaning instead of becoming a decode failure.
fn decode<T: for<'a> serde::Deserialize<'a>>(status: u16, body: &[u8]) -> AppResult<T> {
    if status != 200 {
        let message = serde_json::from_slice::<serde_json::Value>(body)
            .ok()
            .and_then(|value| value["message"].as_str().map(str::to_owned))
            .unwrap_or_else(|| "The execution host refused the request".to_owned());
        return Err(match status {
            400 => AppError::BadRequest(message),
            403 => AppError::Forbidden(message),
            404 => AppError::NotFound(message),
            409 => AppError::Conflict(message),
            501 => AppError::Unsupported(message),
            _ => AppError::Internal(message),
        });
    }
    serde_json::from_slice(body).map_err(|_| {
        AppError::Internal("The execution host answered with an unreadable body".into())
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    fn version(path: &str, sha: Option<&str>) -> FileVersion {
        FileVersion {
            path: path.to_owned(),
            exists: sha.is_some(),
            sha256: sha.map(str::to_owned),
            size: sha.map(|_| 3),
            mtime: None,
            ..FileVersion::default()
        }
    }

    fn seed(workspace_id: &str, path: &str, known: FileVersion) {
        let mut registry = registry().unwrap();
        let entry = registry
            .workspaces
            .entry(workspace_id.to_owned())
            .or_insert_with(|| WorkspaceWatch {
                root_path: "/remote/project".into(),
                files: HashMap::new(),
                polling: true,
            });
        entry.files.insert(
            path.to_owned(),
            Watched {
                viewers: HashSet::from(["node".to_owned()]),
                known,
            },
        );
    }

    #[test]
    fn an_identical_rewrite_is_not_reported_as_a_change() {
        let workspace = "remote-watch-identical";
        seed(workspace, "a.txt", version("a.txt", Some("aa")));
        assert!(difference(workspace, vec![version("a.txt", Some("aa"))]).is_empty());
        release_workspace(workspace);
    }

    #[test]
    fn edits_removals_and_reappearances_each_get_their_own_kind() {
        let workspace = "remote-watch-kinds";
        seed(workspace, "a.txt", version("a.txt", Some("aa")));
        let changes = difference(workspace, vec![version("a.txt", Some("bb"))]);
        assert!(matches!(
            changes.as_slice(),
            [WorkspaceEvent::FileChanged {
                kind: FileChangeKind::Modified,
                ..
            }]
        ));
        let changes = difference(workspace, vec![version("a.txt", None)]);
        assert!(matches!(
            changes.as_slice(),
            [WorkspaceEvent::FileChanged {
                kind: FileChangeKind::Removed,
                ..
            }]
        ));
        let changes = difference(workspace, vec![version("a.txt", Some("cc"))]);
        assert!(matches!(
            changes.as_slice(),
            [WorkspaceEvent::FileChanged {
                kind: FileChangeKind::Replaced,
                ..
            }]
        ));
        release_workspace(workspace);
    }

    #[test]
    fn a_path_no_node_watches_any_more_produces_nothing() {
        let workspace = "remote-watch-unwatched";
        seed(workspace, "a.txt", version("a.txt", Some("aa")));
        unregister(workspace, "a.txt", "node").unwrap();
        assert!(difference(workspace, vec![version("a.txt", Some("bb"))]).is_empty());
        release_workspace(workspace);
    }

    #[test]
    fn a_remote_error_keeps_its_status_instead_of_becoming_a_decode_failure() {
        let error =
            decode::<WatchPollResult>(403, br#"{"code":"forbidden","message":"no"}"#).unwrap_err();
        assert!(matches!(error, AppError::Forbidden(message) if message == "no"));
    }
}

File: apps/desktop/src/core/remote/watch.ts
/**
 * 远端工作空间的文件监听：Worker 那边的平台 watcher 推变化，推不了时控制端轮询。
 *
 * 首选**推送**：登记的文件整组交给 Worker（`files.watch`，按工作空间一组），
 * Worker 看着父目录，内容变了就推一帧 `files.changed`（`watch-worker.ts`），这里
 * 与上次看到的比较后发与本机相同的 `file.changed`。注册的答复写 `mode: events`。
 *
 * **轮询**是退路，两种情况用它：
 *
 *  * Worker 不认 `files.watch`（更旧的构建）——每 {@link POLL_MS} 批量问一次版本，
 *    答复写 `mode: poll`，编辑器显示「轮询」徽标；
 *  * 连接断了——推送的前提是连接活着，而连接只在有请求时重建。断线后先退回轮询，
 *    轮询的请求把连接拉起来，握手成功时（`connected`）整组重登并比对一次版本，
 *    补上断线期间错过的变化，然后停掉轮询。
 *
 * 一次读取失败（主机断开、Worker 重启）不算文件被删：什么都不发，等下一次。
 */

import type { WorkspaceEvent } from "../bus";
import type { FileVersion, WatchRegistration } from "../files/watch";
import { workspaceRelativePath } from "../workspaces/roots";
import { badRequest } from "../workspaces/support";
import { type ExecutionTarget, executeOn, listenRemote } from "./execute";

/** 两次询问之间的间隔。 */
export const POLL_MS = 2_000;

type Publisher = (workspaceId: string, event: WorkspaceEvent) => void;

interface WatchedFile {
  readonly viewers: Set<string>;
  known: FileVersion;
}

interface WorkspaceWatch {
  target: ExecutionTarget;
  publish: Publisher;
  readonly files: Map<string, WatchedFile>;
  timer: NodeJS.Timeout | undefined;
  polling: boolean;
  /** `events`：Worker 在推；`poll`：这里在问。 */
  mode: "events" | "poll";
  /** 一次整组重登在路上时，后来的改动只记一笔，等它回来再发一次。 */
  syncing: Promise<void> | undefined;
  resync: boolean;
}

function hostOf(target: ExecutionTarget): string {
  return target.executionHostId ?? "";
}

export class RemoteWatches {
  private readonly watches = new Map<string, WorkspaceWatch>();
  private readonly unlisten: () => void;

  constructor(private readonly intervalMs: number = POLL_MS) {
    this.unlisten = listenRemote({
      event: (hostId, channel, event) => {
        if (channel !== "control" || event.type !== "files.changed") return;
        const workspaceId =
          typeof event.watchId === "string" ? event.watchId : "";
        const entry = this.watches.get(workspaceId);
        if (entry === undefined || hostOf(entry.target) !== hostId) return;
        if (typeof event.path !== "string") return;
        const version = event.version as FileVersion | undefined;
        if (version === undefined) return;
        this.observe(workspaceId, entry, event.path, version);
      },
      connected: (hostId, channel) => {
        if (channel !== "control") return;
        for (const [workspaceId, entry] of this.watches) {
          if (hostOf(entry.target) === hostId) {
            void this.sync(workspaceId, entry);
          }
        }
      },
      disconnected: (hostId, channel) => {
        if (channel !== "control") return;
        for (const [workspaceId, entry] of this.watches) {
          if (hostOf(entry.target) !== hostId) continue;
          entry.mode = "poll";
          this.arm(workspaceId, entry);
        }
      },
    });
  }

  async register(
    target: ExecutionTarget,
    workspaceId: string,
    requested: string,
    nodeId: string,
    publish: Publisher,
  ): Promise<WatchRegistration> {
    if (nodeId === "" || nodeId.length > 128) {
      throw badRequest("A node id is required");
    }
    const relative = workspaceRelativePath(requested);
    const version = (await executeOn(target, "files.version", {
      path: relative,
    })) as FileVersion;
    let entry = this.watches.get(workspaceId);
    if (entry === undefined) {
      entry = {
        target,
        publish,
        files: new Map(),
        timer: undefined,
        polling: false,
        mode: "poll",
        syncing: undefined,
        resync: false,
      };
      this.watches.set(workspaceId, entry);
    }
    // 换了主机或根，之前记下的版本都是另一处的，不能拿来比。
    if (
      entry.target.rootPath !== target.rootPath ||
      hostOf(entry.target) !== hostOf(target)
    ) {
      entry.files.clear();
    }
    entry.target = target;
    entry.publish = publish;
    const file = entry.files.get(relative) ?? {
      viewers: new Set<string>(),
      known: version,
    };
    file.viewers.add(nodeId);
    // 重新打开即重新取基线：此刻磁盘上的就是节点显示的。
    file.known = version;
    entry.files.set(relative, file);
    await this.sync(workspaceId, entry);
    return {
      status: "watching",
      mode: entry.mode,
      version,
    };
  }

  unregister(workspaceId: string, requested: string, nodeId: string): void {
    const entry = this.watches.get(workspaceId);
    if (entry === undefined) return;
    let relative: string;
    try {
      relative = workspaceRelativePath(requested);
    } catch {
      return;
    }
    const file = entry.files.get(relative);
    if (file === undefined) return;
    file.viewers.delete(nodeId);
    if (file.viewers.size === 0) entry.files.delete(relative);
    if (entry.files.size === 0) this.releaseWorkspace(workspaceId);
    else if (entry.mode === "events") void this.sync(workspaceId, entry);
  }

  releaseWorkspace(workspaceId: string): void {
    const entry = this.watches.get(workspaceId);
    if (entry === undefined) return;
    if (entry.timer !== undefined) clearInterval(entry.timer);
    entry.timer = undefined;
    this.watches.delete(workspaceId);
    if (entry.mode === "events") {
      // 尽力而为：连接已经断了的话，那边的 watcher 已随 Worker 一起没了。
      void executeOn(entry.target, "files.unwatch", {
        watchId: workspaceId,
      }).catch(() => undefined);
    }
  }

  watchedPaths(workspaceId: string): string[] {
    return [...(this.watches.get(workspaceId)?.files.keys() ?? [])];
  }

  /** 这个工作空间现在是推送还是轮询；没有登记是 `undefined`。给测试与诊断。 */
  modeOf(workspaceId: string): "events" | "poll" | undefined {
    return this.watches.get(workspaceId)?.mode;
  }

  shutdown(): void {
    for (const id of [...this.watches.keys()]) this.releaseWorkspace(id);
  }

  /** 停止订阅远端事件；只在测试里造了自己的一份时用。 */
  dispose(): void {
    this.shutdown();
    this.unlisten();
  }

  /**
   * 把整组交给 Worker 并比对它答的当前版本。成功即推送模式、停轮询；Worker 不认
   * 或连接不通就退回轮询。
   */
  private async sync(
    workspaceId: string,
    entry: WorkspaceWatch,
  ): Promise<void> {
    if (entry.syncing !== undefined) {
      entry.resync = true;
      await entry.syncing;
      return;
    }
    entry.syncing = (async () => {
      do {
        entry.resync = false;
        const paths = [...entry.files.keys()];
        if (paths.length === 0) return;
        try {
          const answered = (await executeOn(entry.target, "files.watch", {
            watchId: workspaceId,
            paths,
          })) as { versions?: (FileVersion | null)[] };
          if (this.watches.get(workspaceId) !== entry) return;
          entry.mode = "events";
          if (entry.timer !== undefined) clearInterval(entry.timer);
          entry.timer = undefined;
          paths.forEach((relative, index) => {
            const current = answered.versions?.[index];
            if (current !== undefined && current !== null) {
              this.observe(workspaceId, entry, relative, current);
            }
          });
        } catch {
          if (this.watches.get(workspaceId) !== entry) return;
          entry.mode = "poll";
          this.arm(workspaceId, entry);
          return;
        }
      } while (entry.resync);
    })().finally(() => {
      entry.syncing = undefined;
    });
    await entry.syncing;
  }

  private arm(workspaceId: string, entry: WorkspaceWatch): void {
    if (entry.timer !== undefined) return;
    entry.timer = setInterval(() => {
      void this.poll(workspaceId);
    }, this.intervalMs);
    entry.timer.unref?.();
  }

  /** 与记下的版本比较，变了就发 `file.changed`。 */
  private observe(
    workspaceId: string,
    entry: WorkspaceWatch,
    relative: string,
    current: FileVersion,
  ): void {
    const file = entry.files.get(relative);
    if (file === undefined) return;
    if (
      current.exists === file.known.exists &&
      current.sha256 === file.known.sha256
    ) {
      file.known = current;
      return;
    }
    // 远端答的是版本不是 inode，替换与修改分不开；只有「原来没有」才算替换。
    const kind: "modified" | "removed" | "replaced" = !current.exists
      ? "removed"
      : !file.known.exists
        ? "replaced"
        : "modified";
    file.known = current;
    entry.publish(workspaceId, {
      type: "file.changed",
      workspaceId,
      path: relative,
      kind,
      sha256: current.sha256,
      size: current.size,
      mtime: current.mtime,
    });
  }

  /** 问一轮；上一轮还没回来就跳过，慢主机不会被请求堆满。 */
  async poll(workspaceId: string): Promise<void> {
    const entry = this.watches.get(workspaceId);
    if (entry === undefined || entry.polling) return;
    entry.polling = true;
    try {
      const paths = [...entry.files.keys()];
      if (paths.length === 0) return;
      let versions: (FileVersion | null)[];
      try {
        versions = (await executeOn(entry.target, "files.versions", {
          paths,
        })) as (FileVersion | null)[];
      } catch {
        return;
      }
      if (this.watches.get(workspaceId) !== entry) return;
      paths.forEach((relative, index) => {
        const current = versions[index];
        if (current === undefined || current === null) return;
        this.observe(workspaceId, entry, relative, current);
      });
    } finally {
      entry.polling = false;
    }
  }
}

/** core 里唯一的一份。 */
export const remoteWatches = new RemoteWatches();
